src/Database/PostgreSQL/PQTypes/Internal/State.hs

```diff
@@ -1,49 +1,273 @@
 -- | Definition of internal DBT state.
 module Database.PostgreSQL.PQTypes.Internal.State
-  ( DBState (..)
+  ( -- * ConnectionData
+    ConnectionData
+  , getConnectionSource
+  , getConnectionAcquisitionModeIO
+  , withConnectionData
+  , changeAcquisitionModeTo
+  , withConnection
+
+    -- * DBState
+  , DBState (..)
+  , mkDBState
   , updateStateWith
   ) where
 
+import Control.Concurrent.MVar.Lifted
+import Control.Monad
+import Control.Monad.Base
+import Control.Monad.Catch
+import Data.Function
+import Data.Typeable
 import Foreign.ForeignPtr
+import GHC.Stack
 
+import Data.Monoid.Utils
 import Database.PostgreSQL.PQTypes.FromRow
+import Database.PostgreSQL.PQTypes.Internal.BackendPid
 import Database.PostgreSQL.PQTypes.Internal.C.Types
 import Database.PostgreSQL.PQTypes.Internal.Connection
+import Database.PostgreSQL.PQTypes.Internal.Exception
 import Database.PostgreSQL.PQTypes.Internal.QueryResult
+import Database.PostgreSQL.PQTypes.SQL
 import Database.PostgreSQL.PQTypes.SQL.Class
 import Database.PostgreSQL.PQTypes.Transaction.Settings
 
```

Comment:
I think this warrants some documentation, as it's quite a bit to wrap my head around. Perhaps a name like `LazyConnection` instead of `ConnectionState` to better convey the idea?